simplex/bls.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package simplex
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/simplex"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+)
+
+var (
+	errSignatureVerificationFailed = errors.New("signature verification failed")
+	errSignerNotFound              = errors.New("signer not found in the membership set")
+	simplexLabel                   = []byte("simplex")
+)
+
+var _ simplex.Signer = (*BLSSigner)(nil)
+
+type SignFunc func(msg []byte) (*bls.Signature, error)
+
+// BLSSigner signs messages encoded with the provided ChainID and SubnetID.
+// using the SignBLS function.
+type BLSSigner struct {
+	chainID  ids.ID
+	subnetID ids.ID
+	// signBLS is passed in because we support both software and hardware BLS signing.
+	signBLS SignFunc
+}
+
+type BLSVerifier struct {
+	nodeID2PK map[ids.NodeID]bls.PublicKey
+	subnetID  ids.ID
+	chainID   ids.ID
+}
+
+func NewBLSAuth(config *Config) (BLSSigner, BLSVerifier) {
+	verifier := createVerifier(config)
+
+	return BLSSigner{
+		chainID:  config.Ctx.ChainID,
+		subnetID: config.Ctx.SubnetID,
+		signBLS:  config.SignBLS,
+	}, verifier
+}
+
+// Sign returns a signature on the given message using BLS signature scheme.
+// It encodes the message to sign with the chain ID, and subnet ID,
+func (s *BLSSigner) Sign(message []byte) ([]byte, error) {
+	message2Sign, err := encodeMessageToSign(message, s.chainID, s.subnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode message to sign: %w", err)
+	}
+
+	sig, err := s.signBLS(message2Sign)
+	if err != nil {
+		return nil, err
+	}
+
+	sigBytes := bls.SignatureToBytes(sig)
+	return sigBytes, nil
+}
+
+type encodedSimplexSignedPayload struct {
+	Message  []byte `serialize:"true"`
+	ChainID  []byte `serialize:"true"`
+	SubnetID []byte `serialize:"true"`
+	Label    []byte `serialize:"true"`
```

Comment:
This applies the below suggestions, as well as using `ids.ID` rather than `[]byte` when possible. `ids.ID` is packed more efficiently than `[]byte` (as the length is not included as it is fixed based on the type).
```suggestion
	NetworkID uint32 `serialize:"true"`
	ChainID   ids.ID `serialize:"true"`
	Message   []byte `serialize:"true"`
```